docs/developer/best-practice.md

```diff
@@ -22,258 +41,86 @@ Finally, we have delivery commitments. Larger more intentional refactors may be
+Package level state is rarely required. If it is, it should be composed of re-usable constructs defined and exposed from the package itself.
 
-Factory functions are not to be confused with initializing constructors. Factory functions typically return abstractions (an interface) satisfied by 1 of several implementations. A factory could be as simple as:
+See https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_avoid_package_level_state).
 
-```go
-// Foo and Bar satisfy fmt.Stringer
-func NewStringer(key string) (fmt.Stringer, error) {
-    switch key {
-    case "foo":
-        return Foo{}, nil
-    case "bar":
-        return Bar{}, nil
-    default:
-        return nil, errors.New("unknown key")
-    }
-}
-```
+### "Accept interfaces, return structs"
 
-Factories generally shouldn't contain complex construction logic. That's the job of the initializing constructor for the concrete type.
+This section is included because the phrase surfaces regularly.
 
-##### Loggers
+Code Review Comments provides a simple explanation that covers the general case. However, we want to stress this is the general case, not all cases.
 
-Loggers are sometimes treated as a special dependency by making entire projects depend on a package level logger. This _masks_ a crucial dependency that is hard to change without impacting the rest of the code base. To understand why this is bad, consider extracting a package into its own repository for reusability. Depending on a package level logger, or any other package level dependencies, makes the extraction process unnecessarily difficult.
+When developing abstractions it may become necessary to return interfaces but this should be considered orthogonal to accepting interfaces.
 
-Logger's are not special dependencies, inject them like any other dependency. If you want to minimize the surface area of the default initializing constructor (`New...` funcs) you may construct noop loggers to satisfy the logger dependency and provide a means to override the logger with a setter.
+See https://github.com/golang/go/wiki/CodeReviewComments#interfaces
 
-### Testing
-
-Software development is costly and hazardous. Tests provide example usage, guard rails, and assurances. Without tests, you cannot have high quality code. For this reason, all code we spend time designing and writing is worth testing. Similarly, if it can be unit tested, it should be unit tested. Unit tests are the first line of defense and can be complimented with additional integration or end-to-end tests.
+### Unit tests
 
 Unit tests should:
 
 - focus on the public API of a package.
-- use the `_test` idiom for testing files.
+- prefer the `_test` when testing the public API of the package.
 - treat the package as a black box.
-- cover at least 80% of the public API.
-- named after the entity they're testing; for example, `TestFooBarDependencyErrors` tests that an error occurs when `Foo` is executed and its `Bar` dependency errors.
```

Comment:
I would prefer to keep this
I don't really care what the common format is, but it does help to have one. And we have it (theoretically)

Reply:
Understood. I'm OK keeping it but perhaps tweaking the wording? Naming after the entity under test isn't applicable for functions, for example.

I agree though, clarity and consistency is useful. 

Do you have any ideas on tweaks to make it more general without being so focused on 'entities'?

Reply:
Maybe?

> Test names:
> * If you are testing a function, the test name should start with the name of that function followed by the scenario you are testing; for example, `TestFooInvalidInputError` tests that an error occurs when `Foo` is executed with invalid input.
> * If you are testing a method, the test name should start with the name of the receiver type followed by the method name; for example, `TestFooBarDependencyErrors` tests that an error occurs when `Foo` is executed and its `Bar` errors because of a dependency.